### PR TITLE
Performance optimizations

### PR DESCRIPTION
**Pull request by stefanatwork**
These are performance improvements for compiling larger kernels. I have tested these with my DPC++ kernel with tens of thousands of variables. For my test case, it reduced ocloc run time from 249s to 166s. I don't know if those changes are beneficial only for my special case or useful for the general audience.

### CODE REVIEW

File: visa/GraphColor.cpp [hunk 1]
```diff
@@ -4849,8 +4855,8 @@ void Augmentation::addSIMDIntfForRetDclares(G4_Declare* newDcl)
     if (dclIt == retDeclares.end())
     {
         MaskDeclares newMask;
-        newMask.first.resize(liveAnalysis.getNumSelectedGlobalVar());
-        newMask.second.resize(liveAnalysis.getNumSelectedGlobalVar());
+        newMask.first.reserve(liveAnalysis.getNumSelectedGlobalVar()/32);
```
Comment (pratikashar) [inline visa/GraphColor.cpp:4858]: why divide by 32?

File: visa/BitSet.cpp [hunk 2]
```diff
@@ -109,27 +112,57 @@ void BitSet::invert(void)
 }
 
 template <typename T>
-void vector_and(T *__restrict__ p1, const T *const p2, unsigned n)
+void vector_and(T *__restrict__ p1, const T *const p2, int n)
 {
-    for (unsigned i = 0; i < n; ++i)
+    int i = 0;
+#ifdef __SSE2__
```
Comment (pratikashar) [inline visa/BitSet.cpp:118]: is there some other way to make VC/gcc vectorize this loop? i'm not sure introducing these intrinsics manually in code base is a good idea. also, do we know how much improvement this particular change gets us?

File: visa/GraphColor.cpp (hunk 1 above)
Comment (stefanatwork) [reply visa/GraphColor.cpp:4858]: This was an ad-hoc decision to keep the memory footprint the same as before. We could try not calling reserve() at all and letting it resize on the fly as necessary, but that may come back with a performance penalty again.

File: visa/BitSet.cpp (hunk 2 above)
Comment (stefanatwork) [reply visa/BitSet.cpp:118]: For gcc, we can tell it to do vectorize via `__attribute__((optimize("tree-vectorize")))`. According to godbolt, MSVC and clang automatically use SSE2 in this loop when compiled with O2.

**Comment by stefanatwork** [comment]
The PR has now been updated to be in sync with the latest master.
Vectorization of BitSet is now left to GCC and not done with Intrinsics.
MaskDeclares is now allocated on the fly and not using std::vector::reserve any more.

Performance: 207s->146s

**Review by pratikashar — APPROVED** [review]

**Comment by pratikashar** [comment]
@stefanatwork  - this change is based on master. we checkin to comp/igc and then the change gets promoted to master. please double check whether the change is based on intended branch.

**Comment by pratikashar** [comment]
Approved for merge

**Comment by pszymich** [comment]
Thanks @stefanatwork and @pratikashar for review. Before a merge, we need to have the patch verified. I'll update as soon as it's done.

**Comment by eero-t** [comment]
> Thanks @stefanatwork and @pratikashar for review. Before a merge, we need to have the patch verified. I'll update as soon as it's done.

@pszymich That was half a year ago.  Any updates?

**Comment by stefanatwork** [comment]
I believe this PR is obsolete now. The addition of sparse bit sets has mostly addressed what this patch as intended to do.

**Pull request closed without merging**